File: solutions/21RelativeSortArray/java/RelativeSortArray.java
import java.util.Arrays;

class RelativeSortArray {
  public int[] relativeSortArray(int[] arr1, int[] arr2) {
    int min = arr1[0], max = arr1[0], n = arr1.length;
    for (int i = 0; i < n; i++) {
      if (arr1[i] < min) {
        min = arr1[i];
        continue;
      }
      if (arr1[i] > max) {
        max = arr1[i];
        continue;
      }
    }

    int[] counts = new int[max - min + 1];
    Arrays.fill(counts, 0);

    for (int num : arr1) {
      counts[num - min]++;
    }

    int[] relativeSort = new int[n];
    int index = 0;

    for (int num : arr2) {
      while (counts[num - min] > 0) {
        relativeSort[index++] = num;
        counts[num - min]--;
      }
    }
    for (int i = 0; i < max - min + 1; i++) {
      while (counts[i] > 0) {
        int num = min + i;
        relativeSort[index++] = num;
        counts[i]--;
      }
    }
    return relativeSort;
  }
}

File: solutions/21RelativeSortArray/python/RelativeSortArray.py
from typing import List


class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        arr_max, arr_min = max(arr1), min(arr1)
        counts = [0 for _ in range(arr_max - arr_min + 1)]

        for num in arr1:
            counts[num - arr_min] += 1

        relativeSort = []
        for num in arr2:
            while counts[num - arr_min] > 0:
                relativeSort.append(num)
                counts[num - arr_min] -= 1
        for i in range(arr_max - arr_min + 1):
            while counts[i] > 0:
                relativeSort.append(arr_min + i)
                counts[i] -= 1
        return relativeSort
